src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
+        private async Task Worker(CancellationToken token)
         {
-            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged)
+            while (!token.IsCancellationRequested)
             {
-                var newDocument = changeEvent.NewSolution.GetDocument(changeEvent.DocumentId);
-
-                this.EmitDiagnostics(newDocument.FilePath);
-                foreach (var document in _workspace.GetOpenDocumentIds().Select(x => _workspace.CurrentSolution.GetDocument(x)))
+                try
+                {
+                    var currentWork = GetThrottledWork();
+                    await Task.WhenAll(currentWork.Select(x => Analyze(x.project, x.workReadySource, token)));
+                    await Task.Delay(100, token);
+                }
+                catch (Exception ex)
                 {
-                    this.EmitDiagnostics(document.FilePath);
+                    _logger.LogError($"Analyzer worker failed: {ex}");
                 }
             }
         }
 
-        public void QueueDiagnostics(params string[] documents)
+        private IEnumerable<(Project project, CancellationTokenSource workReadySource)> GetThrottledWork()
         {
-            this.EmitDiagnostics(documents);
-        }
-
-        private void EmitDiagnostics(params string[] documents)
-        {
-            if (_forwarder.IsEnabled)
+            lock (_workQueue)
             {
-                foreach (var document in documents)
-                {
-                    if (!_openDocuments.Contains(document))
-                    {
-                        _openDocuments.Enqueue(document);
-                    }
-                }
+                var currentWork = _workQueue
+                    .Where(x => x.Value.modified.AddMilliseconds(_throttlingMs) < DateTime.UtcNow)
+                    .OrderByDescending(x => x.Value.modified) // If you currently edit project X you want it will be highest priority and contains always latest possible analysis.
+                    .Take(2) // Limit mount of work executed by once. This is needed on large solution...
+                    .ToList();
 
-                if (!_queueRunning && !_openDocuments.IsEmpty)
+                foreach (var workKey in currentWork.Select(x => x.Key))
                 {
-                    this.ProcessQueue();
+                    _workQueue.TryRemove(workKey, out _);
                 }
+
+                return currentWork.Select(x => (x.Value.project, x.Value.workReadySource));
             }
         }
 
-        private void ProcessQueue()
+        public async Task<IEnumerable<(string projectName, Diagnostic diagnostic)>> GetCurrentDiagnosticResult(IEnumerable<ProjectId> projectIds)
+        {
+            await WaitForInitialStartupWorkIfAny();
+
+            var pendingWork = WaitForPendingWorkIfNeededAndGetIt(projectIds);
+
+            await Task.WhenAll(pendingWork);
```

Comment:
why are these wait ? is the cancellation raised when result is ready or something?

Reply:
how about approach of setting cancellation token to be raised after specific timeout and give that cancellation token to analysis? that seems simpler?